[Filters] Limit the memory cost of FilterResults to 100MB

shallawa opened the pull request:
#### 93ee37d3063287725be2d54acbaf38fade9dd4d5
<pre>
[Filters] Limit the memory cost of FilterResults to 100MB
<a href="https://bugs.webkit.org/show_bug.cgi?id=245230">https://bugs.webkit.org/show_bug.cgi?id=245230</a>

Reviewed by Cameron McCormack.

The SVGFilter is completely discarded if the number of unique effects is more
than 100 or the number of children of any effect is more than 200. This is a
naive way to limit the memory used for caching the results because it does not
consider the size of the FilterImage. Moreover this limitation prevents displaying
the sourceImage itself. So we end up with a blank rectangle.

A better approach is to allow caching the FilterImages up to 100MB per filter.
All FilterImages which are generated after that limit will be discarded after
they are used. Any subsequent filter applying will regenerate these FilterImages.

* LayoutTests/svg/filters/repeated-drop-shadow-effects-expected.svg: Added.
* LayoutTests/svg/filters/repeated-drop-shadow-effects.svg: Added.
* Source/WebCore/platform/graphics/coreimage/FilterImageCoreImage.mm:
(WebCore::FilterImage::memoryCostOfCIImage const):
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::memoryCost const):
* Source/WebCore/platform/graphics/filters/FilterImage.h:
* Source/WebCore/platform/graphics/filters/FilterResults.cpp:
(WebCore::FilterResults::memoryCost const):
(WebCore::FilterResults::canCacheResult const):
(WebCore::FilterResults::setEffectResult):
* Source/WebCore/platform/graphics/filters/FilterResults.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::buildExpression):

Canonical link: <a href="https://commits.webkit.org/264807@main">https://commits.webkit.org/264807@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9566658832da91c38dd7578b9b4e49f035467218

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8653 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8943 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9158 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10309 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8665 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8662 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8915 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11533 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8799 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9823 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10466 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7122 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7916 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15442 "1 flakes 150 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8242 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8064 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11406 "8 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8545 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6973 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7815 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2108 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12027 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8286 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/filters/FilterImage.cpp:
```diff
@@ -87,6 +87,28 @@ FloatPoint FilterImage::mappedAbsolutePoint(const FloatPoint& point) const
+        memoryCost += memoryCostOfCIImage();
+#endif
+
+    RELEASE_ASSERT(!memoryCost.hasOverflowed());
```
Is this needed? I think `Checked`'s `size_t` conversion operator will crash for you. (Although I guess it won't have a useful assertion message?)

heycam commented on Source/WebCore/platform/graphics/filters/FilterResults.cpp:
```diff
@@ -40,8 +42,31 @@ FilterImage* FilterResults::effectResult(FilterEffect& effect) const
+
+    for (auto& result : m_results.values()) {
+        memoryCost += result->memoryCost();
+        RELEASE_ASSERT(!memoryCost.hasOverflowed());
```
And here.

heycam commented on Source/WebCore/platform/graphics/filters/FilterResults.cpp:
```diff
@@ -40,8 +42,31 @@ FilterImage* FilterResults::effectResult(FilterEffect& effect) const
+    return memoryCost;
+}
+
+bool FilterResults::canCacheResult(FilterImage& result) const
```
`const FilterImage& result`

heycam commented on Source/WebCore/platform/graphics/filters/FilterResults.cpp:
```diff
@@ -40,8 +42,31 @@ FilterImage* FilterResults::effectResult(FilterEffect& effect) const
+
+bool FilterResults::canCacheResult(FilterImage& result) const
+{
+    CheckedSize totalMemoryCost = memoryCost() + result.memoryCost();
```
The addition could overflow before it is assigned to the `Checked`. Instead maybe:

```
CheckedSize totalMemoryCost;
totalMemoryCost += memoryCost();
totalMemoryCost += result.memoryCost();
```

shallawa replied:
Fixed.

heycam commented on Source/WebCore/platform/graphics/filters/FilterResults.cpp:
```diff
@@ -40,8 +42,31 @@ FilterImage* FilterResults::effectResult(FilterEffect& effect) const
+bool FilterResults::canCacheResult(FilterImage& result) const
+{
+    CheckedSize totalMemoryCost = memoryCost() + result.memoryCost();
+    if (totalMemoryCost.hasOverflowed())
+        return false;
```
Earlier we will crash if the two individual `memoryCost()` calls overflow while they're computing their costs. Do you want to do the same here too, rather than return false?

shallawa replied:
I think this different from `memoryCost()`. FilterImage and FilterResults should stay always within limits. So every time we call `memoryCost()` we can assert/crash we do not overflow. But here we are checking whether adding a new FilterImage to FilterResults will make us within the limits or not.

heycam commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -33,8 +33,7 @@
 namespace WebCore {
 
-static constexpr unsigned maxTotalNumberFilterEffects = 100;
-static constexpr unsigned maxCountChildNodes = 200;
```
Can you explain this part of the change?

shallawa replied:
Set maxCountChildNodes back to 200 and removing maxTotalNumberFilterEffects was explained in the commit message.

heycam commented on Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
```diff
@@ -233,15 +232,14 @@ OptionSet<FilterRenderingMode> SVGFilter::supportedFilterRenderingModes() const
     }
 
-    return copyToVector(effects);
+    return effects;
```
Seems unrelated to the purpose of this bug, so should probably be in a separate PR.

shallawa replied:
Okay. Removed.

shallawa replied:
It is now addressed in https://github.com/WebKit/WebKit/pull/14596

heycam reviewed (commented):
Looks good but I'd like to understand no longer checking `maxTotalNumberFilterEffects`.

shallawa committed:
[Filters] Limit the memory cost of FilterResults to 100MB
https://bugs.webkit.org/show_bug.cgi?id=245230

Reviewed by Cameron McCormack.

The SVGFilter is completely discarded if the number of unique effects is more
than 100 or the number of children of any effect is more than 200. This is a
naive way to limit the memory used for caching the results because it does not
consider the size of the FilterImage. Moreover this limitation prevents displaying
the sourceImage itself. So we end up with a blank rectangle.

A better approach is to allow caching the FilterImages up to 100MB per filter.
All FilterImages which are generated after that limit will be discarded after
they are used. Any subsequent filter applying will regenerate these FilterImages.

* LayoutTests/svg/filters/repeated-drop-shadow-effects-expected.svg: Added.
* LayoutTests/svg/filters/repeated-drop-shadow-effects.svg: Added.
* Source/WebCore/platform/graphics/coreimage/FilterImageCoreImage.mm:
(WebCore::FilterImage::memoryCostOfCIImage const):
* Source/WebCore/platform/graphics/filters/FilterImage.cpp:
(WebCore::FilterImage::memoryCost const):
* Source/WebCore/platform/graphics/filters/FilterImage.h:
* Source/WebCore/platform/graphics/filters/FilterResults.cpp:
(WebCore::FilterResults::memoryCost const):
(WebCore::FilterResults::canCacheResult const):
(WebCore::FilterResults::setEffectResult):
* Source/WebCore/platform/graphics/filters/FilterResults.h:
* Source/WebCore/svg/graphics/filters/SVGFilter.cpp:
(WebCore::SVGFilter::buildExpression):

Canonical link: https://commits.webkit.org/264807@main
